philander/stc311.py

```diff
@@ -0,0 +1,665 @@
+    #  */
+    def _transferCurrentAvg(self, data):  # STC3117 exclusive
+        if self.REGISTER.CHIP_TYPE is not ChipType.STC3117:
+            ret = Current.invalid  # Function is not implemented for this chip
+        else:
+            # Again, we actually read out the voltage drop over the sense resistor.
+            # LSB is 1.47V, partial scaling factor is 1.47 = 147/100
+            # Value is signed!
+            # Total scaling factor is: 147 / 100 * 1000 / rs = 147 * 10 / rs = 1470 / rs.
+            rs = self.REGISTER.CONFIG_GASGAUGE_0_RSENSE
+            if data >= 0:
+                ret = (data * 1470 + rs/2) / rs
+            else:
+                ret = (data * 1470 - rs/2) / rs
+            ret = Current(ret)
+        return ret
+
+    @staticmethod
+    def _transferTemperature(data):
+        # LSB is 1C, so we don't need any scaling.
+        return Temperature(data)
+
+    @staticmethod
+    def _transfertOCV(data):
+        # LSB is 0.55 mV, so scale by factor 0.55 = 55/100 = 11/20.
+        ret = (data * 11 + 10) / 20
+        return Voltage(ret)
+
+    @staticmethod
+    def _invTransferOCV(value):
+        # LSB is 0.55 mV, so scale by factor 1/0.55 = 100/55 = 20/11.
+        ret = (value * 20 + 5) / 11
+        return ret
+
+    @staticmethod
+    def _crc(data, length):
+        ret = 0
+        for idx in range(length):
+            ret ^= data[idx]
+        return ret
+
+    @staticmethod
+    def _checkRamConsistency(self, data, length):
+        # TODO: why is length never used? Refer to original implementation to see, what it does
+        # check if RAM test register value is correct
+        if data[self.REGISTER.IDX_RAM_TEST] != self.REGISTER.RAM_TEST:
+            ret = ErrorCode.errCorruptData
+        else:
+            code = self._crc(data, self.REGISTER.RAM_SIZE - 1)
+            if code != data[self.REGISTER.IDX_RAM_CRC]:
+                ret = ErrorCode.errCorruptData
+            else:
+                ret = ErrorCode.errOk
+        return ret
+
+    def _getOperatingMode(self):
+        err, data = SerialBusDevice.readByteRegister(self,  self.REGISTER.REG_MODE)
+        if err == ErrorCode.errOk:
+            if data and self.REGISTER.MODE_GG_RUN:
+                if data and self.REGISTER.MODE_VMODE:
+                    ret = OperatingMode.opModeVoltage
+                else:
+                    ret = OperatingMode.opModeMixed
+            else:
+                ret = OperatingMode.opModeStandby
+        else:
+            ret = OperatingMode.opModeUnknown
+        return ret
+
+    def _checkID(self):
+        data, err = SerialBusDevice.readByteRegister(self, self.REGISTER.REG_ID)
+        if err == ErrorCode.errOk:
+            err = ErrorCode.errOk if (data == self.REGISTER.CHIP_ID) else ErrorCode.errFailure
+        return err
+
+    #
+    # /**
+    #  * Start-up the chip and fill/restore configuration registers
+    #  */
+    def _initialize(self):
```

Comment:
Maybe, we can give it another name? Such as bootUp?

Reply:
Called it setup for now. As mentioned above, because of how it is used, it could also be part of open. But that also might bloat open a lot.